### PR TITLE
Add valid_move? for queen

### PR DESCRIPTION
**Pull request by kellylougheed**
- Wrote logic to validate move for queen (horizontally, vertically, and diagonally) in queen.rb.

- Wrote tests that 1) returned false for an invalid move, 2) returned true for a valid move, 3) returned true if the destination was occupied by a piece of the opposing color, 4) returned false if the destination was occupied by a piece of the same color, and 5) returned false if the queen was obstructed.

- Further refactored is_obstructed? in piece.rb within an inch of its life to ensure that the ranges were looking at ONLY the squares in between the current location and destination, and NOT any square in between. Also had to resolve the additional problem of a range consisting of only one number (e.g. 3...3) returning nil.

All tests are passing!

### CODE REVIEW

**Comment by PamBWillenz** [comment]
Kelly - really impressive logic/code you've written. Lots of detail. The min, max and incrementer are super detailed. I think it is great you can follow the logic of all the conditions you set up. It makes sense to me. I can understand the outcome of what you're checking for. For your queen's vaild_move, did you pick the (4, 4) in the queen factory arbitrarily or is there a reason for those particular coordinates? Nice job on putting these complicated moves into code.

**Comment by kellylougheed** [comment]
Thanks, @pam! I'm glad the logic makes sense (though I've been wondering if should refactor that lengthy code). I put (4,4) in the factory so that the queen is at the center of the board and can move in any direction, but those coordinates can be overridden.

File: app/models/piece.rb [hunk 1]
```diff
@@ -60,7 +66,13 @@ def is_obstructed?(destination_x, destination_y)
       location_x > destination_x ? incrementer = -1 : incrementer = 1
       position_x = location_x + incrementer
       min, max = [position_x, destination_x].minmax
-      range = min..max
+      range = nil
```
Comment (lasley) [inline app/models/piece.rb]: Duplicated code should be a helper method

File: app/models/piece.rb [hunk 2]
```diff
@@ -72,9 +84,25 @@ def is_obstructed?(destination_x, destination_y)
...
       min_x, max_x = [position_x, destination_x].minmax
-      range_x = min_x..max_x
+      range_x = nil
+      if min_x != max_x
+        if x_incrementer > 0
+          max_x = max_x - x_incrementer
+        else
+          min_x = min_x - x_incrementer
+        end
+        min_x == max_x ? range_x = min_x : range_x = min_x...max_x
```
Comment (lasley) [inline app/models/piece.rb]: Copy paste alert!

File: app/models/queen.rb [hunk 3]
```diff
@@ -1,4 +1,15 @@
 class Queen < Piece
+
+  def valid_move?(destination_x, destination_y)
+    valid = super(destination_x, destination_y)
+    horizontal = (destination_y == self.y_position)
```
Comment (lasley) [inline app/models/queen.rb]: Horizontal, vertical, and diag is implemented in https://github.com/the-cheaters/up-all-knight/blob/master/app/models/piece.rb#L48 - maybe make a helper. Although it would be best to submit just the helper in another PR so that team is aware and can use in their moves too without having to wait on this merge.

**Comment by kellylougheed** [comment]
Thanks for the feedback @lasley -- I knew I needed to refactor that code, and now I've done it! It's passing all tests, but I did get a NoMethodError when I tried to explicitly declare "range" a helper method. Is there a different way to do it than simply declaring "helper_method :range"? There are lots of conflicting answers in Google.

**Comment by lasley** [comment]
I guess my wording conflicted with Rails wording. There are Rails specific helper methods, which are for views, and not what we want.

I was going for the comp sci term, which is simply a method to help you do something. It's just a simple def in your model, likely in private scope given this contex 😉

File: app/models/queen.rb [hunk 4]
```diff
@@ -1,4 +1,15 @@
...
+    vertical = (destination_x == self.x_position)
+    diagonal = (
+      (destination_x - self.x_position).abs == 
+      (destination_y - self.y_position).abs
+    )
+    valid && (horizontal || vertical || diagonal)
```
Comment (FsDevNinja) [inline app/models/queen.rb]: so it looks like your queen file needs to be updated with the latest helper methods that are in the piece model.

Comment (kellylougheed) [inline app/models/queen.rb]: Changes made!
